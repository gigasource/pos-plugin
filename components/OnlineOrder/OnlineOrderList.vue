<template>
  <div class="online-order-list">
    <div class="row-flex align-items-center mb-2">
      <div class="online-order-list__title">
        {{`${$t(`onlineOrder.${status}`)} ${$t('onlineOrder.orders')}`}}
      </div>
      <g-spacer/>
      <span class="text-grey-darken-1 fs-small mr-1">{{$t('onlineOrder.orders')}}:</span>
      <span class="fw-700 fs-large mr-3">{{ totalOrder }}</span>
      <span class="text-grey-darken-1 fs-small mr-1">{{$t('onlineOrder.total')}}:</span>
      <span class="fw-700 fs-large mr-3">{{$t('common.currency', storeLocale)}}{{ totalIncome | formatMoney }}</span>
      <date-range-picker :from="filter.fromDate" :to="filter.toDate" @save="changeFilter"/>
    </div>
    <div class="online-order-list__table">
      <g-table elevation="2" fixed-header>
        <thead>
        <tr>
          <th v-for="header in headers">{{header}}</th>
        </tr>
        </thead>
        <tbody>
        <tr v-for="(item, i) in computedItems" :key="i">
          <td class="fw-700">
            <p>#{{item.id}}</p>
            <g-tooltip :open-on-hover="true" color="#616161" transition="0.3" speech-bubble remove-content-on-close>
              <span><b>From:</b> {{item.forwardedStore}}</span>
              <template v-slot:activator="{on}">
                <div v-on="on" v-if="item.forwardedStore">
                  <g-icon size="16">icon-double-arrow-right_blue</g-icon>
                </div>
              </template>
            </g-tooltip>
          </td>
          <td>
            <p>{{item.customer.name}}</p>
            <p>{{item.customer.phone}}</p>
          </td>
          <td>
            <div v-if="item.customer.address">
              <p>{{item.customer.address}}</p>
              <p>{{item.customer.zipCode}}</p>
            </div>
            <div v-else>--</div>
          </td>
          <td>
            <p class="fw-700">{{$t('common.currency', storeLocale)}}{{item.payment[0].value | formatMoney}}</p>
            <p>{{item.payment[0].type}}</p>
          </td>
          <td>{{item.date | formatDate}}</td>
          <td>{{item.deliveryTime}}</td>
          <td class="fw-700">{{$t(`onlineOrder.${item.type}`)}}</td>
          <td :class="statusClass">
            <div>{{$t(`onlineOrder.${item.status}`)}}</div>
            <div style="font-size: x-small; margin-top: -5px"> {{ isRefunded(item) ? refundedStr: '' }}</div>
          </td>
          <td>
            <!-- ATM, g-menu only contains 1 action so we set this action state as visibility of entire g-menu -->
            <g-menu v-if="isAdminUser && isRefundable(item)" v-model="item.showMenu" :nudge-bottom="10" close-on-content-click>
              <template #activator="{on}">
                <div @click="on.click" :style="moreMenuStyle">···</div>
              </template>
              <template #default>
                <g-card background="white">
                  <div style="padding: 10px; cursor: pointer" @click="$emit('refundOrder', item, status)">{{ refundStr }}</div>
                </g-card>
              </template>
            </g-menu>
          </td>
        </tr>
        </tbody>
      </g-table>
    </div>
  </div>
</template>

<script>
  import dayjs from 'dayjs'
  import isBetween from 'dayjs/plugin/isBetween'
  dayjs.extend(isBetween)

  export default {
    name: 'OnlineOrderList',
    components: { },
    props: {
      status: String,
      onlineOrders: {
        type: Array,
        default: () => []
      },
      storeLocale: String,
      user: Object,
      isRefundFailed: Function, // (order) => boolean: return refund status of order
      isRefundable: Function, // (order) => boolean: return true if order can be refund
      isRefunded: Function,
      getOnlineOrdersWithStatus: Function,
    },
    data() {
      const i18n = this.$i18n;
      const { onlineOrder: { address, amount, customer, delivery, no, received, status, type, refund, refunded } } = i18n.messages[i18n.locale] || i18n.messages[i18n.fallbackLocale]
      const headers = [no, customer, address, amount, received, delivery, type, status, '']
      return {
        headers,
        refundStr: refund,
        refundedStr: refunded,
        filter: {
          fromDate: '',
          toDate: ''
        }
      }
    },
    filters: {
      formatDate(date) {
        return dayjs(date).format('HH:mm')
      },
      formatMoney(value, decimals = 2) {
        if (value != null)
          return !isNaN(value) ? value.toFixed(decimals) : value
      }
    },
    computed: {
      computedItems() {
        const items = this.onlineOrders.map(i => ({ ...i, status: this.status }))
        if(this.filter.fromDate && this.filter.toDate) {
          return items.filter(item => dayjs(item.date).isBetween(this.filter.fromDate, this.filter.toDate, 'day', '[]'))
        }
        return items
      },
      statusClass() {
        return this.status
      },
      totalOrder() {
        return this.computedItems.length
      },
      totalIncome() {
        return _.sumBy(this.computedItems, item => item.vSum);
      },
      isAdminUser() {
        return this.user.role === "admin"
      },
      moreMenuStyle() {
        const color = '#999'
        return {
          'text-align': 'center',
          'line-height': '18px',
          color: color,
          // border: `1px solid ${color}`,
          'border-radius': '50%',
          width: '20px',
          height: '20px',
          'box-sizing': 'border-box',
          'box-shadow': '0px 0px 3px 0px rgba(0,0,0, 0.6)',
          cursor: 'pointer',
          'background-color': '#fff'
        }
      }
    },
    watch: {
      async status(val) {
        if (!val) return
        await this.getOnlineOrdersWithStatus(this.status)
      }
    },
    async mounted() {
      if (this.status) await this.getOnlineOrdersWithStatus(this.status)
      this.filter.fromDate = dayjs().format('YYYY-MM-DD')
      this.filter.toDate = dayjs().format('YYYY-MM-DD')
    },
    async activated() {
      if (this.status) await this.getOnlineOrdersWithStatus(this.status)
    },
    methods: {
      changeFilter(range) {
        this.filter = range
      }
    }
  }
</script>

<style scoped lang="scss">
  .online-order-list {
    background-image: url('/plugins/pos-plugin/assets/out.png');
    width: 100%;
    height: 100%;
    padding: 16px 16px 24px 16px;

    &__title {
      font-size: 15px;
      font-weight: 700;
      text-transform: capitalize;
    }

    .g-select ::v-deep {
      .bs-tf-label {
        display: none;
      }

      .bs-tf-inner-input-group {
        padding: 0 0 0 8px;
      }

      .input {
        font-size: 14px;
        min-width: 85px;
        text-align: center;
      }
    }

    &__table {
      height: calc(100% - 40px);
      width: 100%;

      .g-table {
        th, td {
          height: 69px;
          padding: 8px 4px 0;
          vertical-align: top;
          font-size: 14px;
          word-break: break-word;

          &:nth-child(1) {
            width: 6%;
          }

          &:nth-child(2),
          &:nth-child(3) {
            width: 21%;
          }

          &:nth-last-child(-n+6) {
            width: 10%;
          }

          // actions menu
          &:nth-last-child(1) {
            width: 40px;
          }
        }

        thead th {
          background: #EFEFEF;
          font-size: 12px;
          color: #757575;
          height: 38px;
        }

        tr:nth-child(even) {
          td {
            background: #F8F8FB;
          }
        }

        .completed {
          color: #4CAF50;
          text-transform: capitalize;
        }

        .declined {
          color: #FF5252;
          text-transform: capitalize;
        }
      }
    }

    ::v-deep .bs-tf-wrapper {
      margin: 8px 0;
      background-color: #FFF;
      width: 150px;

      .bs-tf-label {
        margin-bottom: 0;
      }
    }
  }
</style>
