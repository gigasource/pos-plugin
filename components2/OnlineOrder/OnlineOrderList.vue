<script>
import { useI18n } from 'vue-i18n';
import dayjs from 'dayjs';
import { user } from '../AppSharedStates';
import { ref, computed, onBeforeMount } from 'vue'
import _ from 'lodash';
import { $filters } from '../AppSharedStates';

export default {
  props: {
    status: String,
    onlineOrders: {
      type: Array,
      default: () => []
    }
  },
  setup() {
    const { t, locale } = useI18n()
    const onlineOrders = ref(props.onlineOrders)
    const status = ref(props.status)
    const dialogRef = ref(null)
    const i18n = useI18n();
    const { onlineOrder: { address, amount, customer, delivery, no, received, _status, type, refund, refunded } } = i18n.messages[i18n.locale] || i18n.messages[i18n.fallbackLocale]
    const headers = [no, customer, address, amount, received, delivery, type, _status, '']
    const refundStr = ref(refund)
    const refundedStr = ref(refunded)
    const filter = ref({
      fromDate: '',
      toDate: ''
    })
    const dialog = ref(false)
    const computedItems = computed(() => {
      const items = onlineOrders.map(i => ({ ...i, status: status }))
      if (filter.value.fromDate && filter.value.toDate) {
        return items.filter(item => dayjs(item.date).isBetween(filter.value.fromDate, filter.value.toDate, 'day', '[]'))
      }
      return []
    })
    const statusClass = computed(() => {
      return props.status
    })
    const totalOrder = computed(() => {
      return computedItems.value.length
    })
    const totalIncome = computed(() => {
      return _.sumBy(computedItems.value, item => item.vSum);
    })
    const isAdminUser = computed(() => {
      return user.value.role === 'admin'
    })
    const moreMenuStyle = {
      'text-align': 'center',
      'line-height': '18px',
      color: '#999',
      // border: `1px solid ${color}`,
      'border-radius': '50%',
      width: '20px',
      height: '20px',
      'box-sizing': 'border-box',
      'box-shadow': '0px 0px 3px 0px rgba(0,0,0, 0.6)',
      cursor: 'pointer',
      'background-color': '#fff'
    }

    function changeFilter(range) {
      filter.value = range
    }

    function getImagePayment(type) {
      let paymentMethod = cms.getList('PosSetting')[0].payment.find(i => i.name === type)
      return paymentMethod && paymentMethod.icon
    }

    function openDialogDetail(item) {
      dialogRef.value.showDialog(item)
    }

    function formatDate(date) {
      if (!date || !dayjs(date).isValid()) return ''
      return dayjs(date).format('HH:mm')
    }

    function isCaptureRefundExpired(captureResponses) {
      // find final capture
      let finalCapture;
      for (let purchase_unit of captureResponses.purchase_units) {
        for (let capture of purchase_unit.payments.captures) {
          if (capture.final_capture) {
            finalCapture = capture
            break;
          }
        }
      }

      // capture not completed, doesn't allow refund
      if (!finalCapture)
        return true;

      const _3HoursBefore = dayjs().subtract(3, 'hour')
      const createTime = dayjs(finalCapture.create_time)
      return createTime.isBefore(_3HoursBefore)
    }
    function isRefundable(order) {
      // refundable order is order paid via paypal and money has been captured
      // and not refund yet or refund but some capture failed
      return (order.paypalOrderDetail
          && order.paypalOrderDetail.captureResponses
          && order.paypalOrderDetail.captureResponses.status === 'COMPLETED' && !isCaptureRefundExpired(order.paypalOrderDetail.captureResponses)
          && (!order.paypalOrderDetail.refundResponses || isRefundFailed(order.paypalOrderDetail.refundResponses)))
    }

    function isRefunded(order) {
      return (order.paypalOrderDetail
          && order.paypalOrderDetail.refundResponses
          && _.every(order.paypalOrderDetail.refundResponses, r => r.status === 'COMPLETED'))
    }

    function isRefundFailed(refundResponses) {
      return _.find(refundResponses, r => r.status !== 'COMPLETED') != null
    }

    async function getOnlineOrdersWithStatus(status) {
      onlineOrders.value = await cms.getModel('Order').find({
        online: true,
        status
      })
    }

    watch(() => props.status, async (newV) => {
      if (!newV) return
      await getOnlineOrdersWithStatus(newV)
    })

    onBeforeMount(async () => {
      if (status.value) await getOnlineOrdersWithStatus(status.value)
      filter.value.fromDate = dayjs().format('YYYY-MM-DD')
      filter.value.toDate = dayjs().format('YYYY-MM-DD')
    })

    return () => <>
      <div class="online-order-list">
        <div class="row-flex align-items-center mb-2">
          <div class="online-order-list__title">
            {`${t(`onlineOrder.${status.value}`)} ${t('onlineOrder.orders')}`}
          </div>
          <g-spacer></g-spacer>
          <div class="online-order-list__info">
            <div class="row-flex align-items-center">
              <g-icon class="mr-2" size="20">
                icon-cutleries2
              </g-icon>
              <span class="fw-700">
                {totalOrder.value} </span>
            </div>
            <div class="row-flex align-items-center">
              <g-icon class="mr-2" size="20">
                icon-money-bag
              </g-icon>
              <span class="fw-700">
                {t('common.currency', locale.value)}{$filters.formatCurrency(totalIncome.value)} </span>
            </div>
          </div>
          <date-range-picker from={filter.value.fromDate} to={filter.value.toDate} onSave={changeFilter}></date-range-picker>
        </div>
        <div class="online-order-list__table">
          <g-table elevation="2" fixed-header>
            <thead>
            <tr>
              {headers.map(header =>
                  <th style="white-space: nowrap">
                    {header} </th>
              )} </tr>
            </thead>
            <tbody>
            {computedItems.value.map((item, i) =>
                <tr key={i} onClick={() => openDialogDetail(item)}>
                  <td class="fw-700">
                    <p style="white-space: nowrap">
                      #{item.dailyId ? item.dailyId : item.id} </p>
                    <g-tooltip openOnHover={true} color="#616161" transition="0.3" speech-bubble remove-content-on-close v-slots={{
                      'default': () => <> <span>
                        <b> From: </b>
                        {item.forwardedStore} </span>
                      </>
                      ,
                      'activator': ({ on }) => <>
                        {
                          (item.forwardedStore) &&
                          <div v-on="on">
                            <g-icon size="16">
                              icon-delivery-forward
                            </g-icon>
                          </div>
                        }
                      </>
                    }}></g-tooltip>
                  </td>
                  <td>
                    <p> {item.customer.name} </p>
                    <p style="white-space: nowrap"> {item.customer.phone} </p>
                  </td>
                  <td>
                    {
                      (item.customer.address) ?
                          <div>
                            <p style="word-break: break-word">
                              {item.customer.address} </p>
                            <p>
                              {item.customer.zipCode} </p>
                          </div>
                          :
                          <div> -- </div>
                    }
                  </td>
                  <td>
                    <p class="fw-700" style="white-space: nowrap">

                      {t('common.currency', locale.value)}{$filters.formatCurrency(item.payment[0].value)}
                    </p>
                    <p>
                      <img alt src={getImagePayment(item.payment[0].type)}> </img>
                    </p>
                  </td>
                  <td style="white-space: nowrap">
                    {formatDate(item.date)} </td>
                  <td style="white-space: nowrap">
                    {item.deliveryTime} </td>
                  <td>
                    {
                      (item.type === 'delivery') &&
                      <g-icon>
                        icon-delivery-scooter </g-icon>
                    }
                    {
                      (item.type === 'pickup') &&
                      <g-icon>
                        icon-take-away </g-icon>
                    }
                  </td>
                  <td class={statusClass.value}>
                    <div style="white-space: nowrap">
                      {t(`onlineOrder.${item.status}`)} </div>
                    <div style="font-size: x-small; margin-top: -5px">
                      {isRefunded(item) ? refundedStr.value : ''} </div>
                  </td>
                  <td>
                    {
                      (isAdminUser.value && isRefundable(item)) &&
                      <g-menu v-model={item.showMenu} nudgeBottom={10} close-on-content-click v-slots={{
                        'default': () => <>
                          <g-card background="white">
                            <div style="padding: 10px; cursor: pointer" onClick={() => emit('refundOrder', item, status)}>
                              {
                                refundStr
                              }
                            </div>
                          </g-card>
                        </>
                        ,
                        'activator': ({ on }) => <>
                          <div onClick={on.click} style={moreMenuStyle}>
                            ···
                          </div>
                        </>
                      }}></g-menu>
                    }
                  </td>
                </tr>
            )} </tbody>
          </g-table>
        </div>
        <dialog-complete-order disabled-btn ref={dialogRef} v-model={dialog.value}></dialog-complete-order>
      </div>
    </>
  }
}
</script>

<style scoped lang="scss">
.online-order-list {
  background-image: url('/plugins/pos-plugin/assets/out.png');
  width: 100%;
  height: 100%;
  padding: 16px 16px 24px 16px;

  &__title {
    font-size: 15px;
    font-weight: 700;
    text-transform: capitalize;
  }

  .g-select ::v-deep {
    .bs-tf-label {
      display: none;
    }

    .bs-tf-inner-input-group {
      padding: 0 0 0 8px;
    }

    .input {
      font-size: 14px;
      min-width: 85px;
      text-align: center;
    }
  }

  &__info {
    display: flex;
    align-items: center;
    margin-right: 16px;
    background-color: #fff;
    border: 1px solid #E0E0E0;
    border-radius: 4px;

    & > div:first-child {
      border-right: 1px solid #E0E0E0;
    }

    div {
      padding: 5px 8px;
    }
  }

  &__table {
    height: calc(100% - 40px);
    width: 100%;

    .g-table {
      th, td {
        height: auto;
        padding: 8px 4px;
        vertical-align: top;
        font-size: 14px;
        word-break: break-word;

        &:nth-child(1) {
          width: 6%;
        }

        &:nth-child(2),
        &:nth-child(3) {
          width: 21%;
        }

        &:nth-last-child(-n+6) {
          width: 10%;
        }

        // actions menu
        &:nth-last-child(1) {
          width: 40px;
        }
      }

      thead th {
        background: #EFEFEF;
        font-size: 12px;
        color: #757575;
        height: 38px;
      }

      tr:nth-child(even) {
        td {
          background: #F8F8FB;
        }
      }

      .completed {
        color: #4CAF50;
        text-transform: capitalize;
      }

      .declined {
        color: #FF5252;
        text-transform: capitalize;
      }
    }
  }

  ::v-deep .bs-tf-wrapper {
    margin: 8px 0;
    background-color: #FFF;
    width: 150px;

    .bs-tf-label {
      margin-bottom: 0;
    }
  }
}

@media screen and (max-width: 1023px) {
  .online-order-list {

    &__table {
      .g-table td {
        font-size: 12px;
      }
    }
  }
}

</style>
