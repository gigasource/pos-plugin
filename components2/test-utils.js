import "../jest.setup";
import _ from 'lodash';
import {config, mount} from "@vue/test-utils";
import {hooks} from "./OrderView/pos-logic";
import {disablePay, payBtnClickable, payPrintMode, showIcon} from "./OrderView/pos-logic-be";
import StubFactory from './StubFactory';

jest.mock("vue-i18n", () => {
  const util = require('util');
  return {
    useI18n() {
      return {
        t(s) {
          if (s === 'dates.dateFormat') return 'DD.MM.YYYY';
          if (s === 'dates.timeFormat') return 'HH:mm';
          if (arguments.length > 1) return [...arguments].map(s => util.inspect(s)).join(' , ')
          return s; // [Consider] return `t('${t}')` to indicate that i18n has been used
        },
        locale: "en"
      };
    }
  };
});
jest.mock("vue-router", () => {
  return {
    useRoute() {
      return {
        query: {
          type: 'default'
        }
      };
    },
    useRouter() {
      return {
        push() {
        }
      };
    }
  };
});

const dayjs = require("dayjs");
export const orm = require("schemahandler");
const syncFlow = require("schemahandler/sync/sync-flow");
const syncPlugin = require("schemahandler/sync/sync-plugin-multi");
const {stringify} = require("schemahandler/utils");
//jest.useFakeTimers("modern").setSystemTime(new Date("2021-01-01").getTime());
require("mockdate").set(new Date("2021-01-01").getTime());

const delay = require("delay");

export const foodTax = {taxes: [5, 10]};
export const drinkTax = {taxes: [16, 32]};

export const cola = {name: "Cola", price: 1.3, quantity: 1, ...drinkTax};
export const fanta = {name: "Fanta", price: 2, quantity: 1, ...drinkTax};
export const rice = {name: "Rice", price: 10, quantity: 1, ...foodTax};
export const ketchup = {name: "Add Ketchup", price: 3, quantity: 1};

beforeAll(async () => {
  //orm.connect({uri: "mongodb://localhost:27017"}, "myproject");
  orm.registerSchema("Order", {
    status: String,
    date: Date,
    vDate: Date,
    items: [{}],
    table: String
  });
  orm("Order").deleteMany().then();
  orm("Commit").deleteMany().then();

  orm.plugin(syncPlugin);
  orm.plugin(syncFlow);
  orm.registerCommitBaseCollection('Order', 'Action');
  orm.plugin(require("../backend/commit/orderCommit"));
  //orm.plugin(require("../backend/commit/actionCommit"));
  orm.registerCollectionOptions("Order");
  orm.emit("commit:flow:setMaster", true);
});

export let actions = [];
["showOrderReceipt", "pay", "printOrder"].forEach(e =>
  hooks.on(e, () => actions.push(e))
);
export let expectArray = () => [
  "payBtnClickable",
  payBtnClickable.value,
  "payPrintMode",
  payPrintMode.value,
  "showIcon",
  showIcon.value,
  "disablePay",
  disablePay.value,
  actions
];
//</editor-fold>

//<editor-fold desc="component">
export let component, wrapper;

export function setComponent(_component) {
  component = _component;
}

/**
 *
 * @param _component
 * @param options
 * @param useDefaults indicate whether _.defaults will be use for options
 * @returns {*}
 * example: makeWrapper(Order2, {shallow: false});
 */
export const makeWrapper = (_component, options, useDefaults = false) => {
  const defaultOption = {
    props: {},
    global: {
      directives: {
        touch: true
      },
      stubs: {
        'choose-table-dialog': true,
        'dialog-change-value': true,
        'dialog-choose-popup-modifier': true,
        "dialog-config-order-item": true,
        'dialog-confirm-delete': true,
        'dialog-form-input': true,
        'dialog-multi-payment': true,
        'dialog-multi-payment2': true,
        'dialog-new-payment': true,
        'dialog-new-tax-category': true,
        'dialog-product-search-result': true,
        'dialog-text-filter': true,
        'g-avatar': true,
        'g-autocomplete': true,
        'g-badge': true,
        'g-btn': true,
        "g-btn-bs": true,
        'g-card': true,
        'g-card-actions': true,
        'g-card-text': true,
        'g-card-title': true,
        'g-combobox': true,
        'g-dialog': true,
        'g-divider': true,
        "g-icon": true,
        'g-item-group': true,
        'g-file-input': true,
        'g-menu': true,
        "g-overlay": true,
        'g-radio': true,
        'g-select': true,
        'g-simple-table': true,
        'g-spacer': true,
        'g-switch': true,
        'g-table': true,
        'g-tabs': true,
        'g-tab-item': true,
        'g-textarea': true,
        'g-text-field': true,
        'g-text-field-bs': true,
        'g-toolbar': true,
        'g-snackbar': true,
        'g-checkbox': true,
        'g-chip': true,
        'g-select': true,
        'pos-file-input-image': true,
        'pos-order': true,
        'pos-quick-order-toolbar': true,
        //'pos-order-split-order': true,
        'new-reservation-dialog': true,
        'pos-keyboard-full': true,
        'pos-order-keyboard': true,
        'pos-order-move-items': true,
        'pos-order-receipt': true,
        'pos-order-voucher-dialog': true,
        'pos-textfield-new': true,
        'pos-time-picker': true,
        'g-grid-select': StubFactory('g-grid-select'),
        'pos-payment-screen-payment-methods2': true,
        'pos-payment-screen-keyboard2': true,
        'pos-restaurant-payment-order-detail2': true,
        'pos-restaurant-payment-toolbar2': true,
        'scroll-select': true,
      },
      mocks: {
        t: a => a,
        $t: a => a,
        $filters: {
          formatCurrency(val, decimals = 2) {
            if (!val || isNaN(val) || Math.floor(val) === val) return val;
            return val.toFixed(decimals);
          }
        }
      }
    }
  }
  const optionResult = useDefaults ? _.defaults(options, defaultOption) : _.merge(defaultOption, options)
  wrapper = mount(component || _component, optionResult);
};

beforeAll(() => (config.renderStubDefaultSlot = true));

import fs from 'fs';
import {PNG} from "pngjs";
global.printFunctions = {
  printPng: async (png) => {
    try {
      const bufferInBase64 = PNG.sync.write(png);
      fs.writeFileSync(__dirname + '/EndOfDay/__test__/__snapshots__/report.png', bufferInBase64);

    } catch (e) {
      console.log('canvasprinter: printPng exception', e)
    }
  },
  print: () => null
}
