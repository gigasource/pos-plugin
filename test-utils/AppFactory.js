export default function (component) {
	return {
		name: 'App',
		components: {
			Component: component
		},
		render() {
			return <Component/>
		}
	}
}
