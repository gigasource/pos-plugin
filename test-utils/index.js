const cmsFactory = require('./mockCms')
const syncFactory = require('./mockSync')

module.exports = {
	cmsFactory,
	syncFactory
}
