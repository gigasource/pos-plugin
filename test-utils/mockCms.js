const { Socket, Io } = require('schemahandler/io/io')
const orm = require('schemahandler/orm')
const Hooks = require('schemahandler/hooks/hooks')
const path = require('path')
const fs = require('fs')
const _ = require('lodash')

function cmsFactory(testName) {
  orm.connect({uri: "mongodb://localhost:27017"}, testName);
  const socketToFrontend = new Io()
  socketToFrontend.listen('frontend')
  const cms = {
    init: async function () {
      await this.initDemoData()
    },
    initDemoData: async function () {
      const dataPath = path.resolve(__dirname, './dataMock/demoData.json')
      const data = JSON.parse(fs.readFileSync(dataPath, 'utf-8'));

      for (const collection in data) {
        await cms.getModel(collection).remove({})
        if (data.hasOwnProperty(collection)) {
          await Promise.all(data[collection].map(async document => {
            await cms.getModel(collection).create(document)
          }))
        }
      }
    },
    orm,
    socket: socketToFrontend,
    getModel: function (modelName) {
      return orm(modelName)
    }
  }
  _.extend(cms, new Hooks())
  return cms
}

module.exports = cmsFactory
