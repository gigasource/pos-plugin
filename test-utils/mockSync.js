const cmsFactory = require('./mockCms')
const syncFlow = require("schemahandler/sync/sync-flow");
const syncPlugin = require("schemahandler/sync/sync-plugin-multi");
const syncTransporter = require('schemahandler/sync/sync-transporter');
const { Socket, Io } = require('schemahandler/io/io')

async function syncFactory(testName, numberOfNode) {
	const cms = []
	for (let i = 0; i < numberOfNode; i++) {
		const _cms = cmsFactory(`${testName}${i}`)
		await _cms.init()
		_cms.orm.plugin(syncPlugin)
		_cms.orm.plugin(syncFlow)
		_cms.orm.plugin(syncTransporter)
		cms.push(_cms)
	}

	const masterIo = new Io()
	masterIo.listen('master')
	masterIo.on('connect', socket => {
		cms[0].orm.emit('initSyncForMaster', socket)
	})
	cms[0].orm.emit('commit:flow:setMaster', true)
	for (let i = 1; i < numberOfNode; i++) {
		const socket = new Socket()
		socket.connect('master')
		cms[i].orm.emit('initSyncForClient', socket)
	}
	return cms
}

module.exports = syncFactory
