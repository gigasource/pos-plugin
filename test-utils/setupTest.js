const { VueLoaderPlugin } = require('vue-loader')
const { WebpackManifestPlugin } = require('webpack-manifest-plugin');
const MiniCssExtractPlugin = require("mini-css-extract-plugin");
const path = require('path')
const webpack = require('webpack')

module.exports = async function (options) {
	const config = {
		entry: options.componentPath,
		output: {
			libraryTarget: 'commonjs2',
			publicPath: '', // prevent prefix 'auto'
			path: path.resolve(__dirname, '../../dist_test')
		},
		resolve: {
			extensions: ['.vue', '.jsx', '.js', '.json'],
			alias: {
				vue: "@vue/runtime-dom",
			},
		},
		externals: {
			vue: 'vue'
		},
		optimization: {
			minimize: false,
		},
		module: {
			rules: [
				{
					test: /\.vue$/,
					use: "vue-loader",
				},
				{
					test: /\.png$/,
					use: {
						loader: "url-loader",
						options: { limit: 8192 },
					},
				},
				{
					test: /\.js?$/,
					loader: "babel-loader"
				},
				{
					test: /\.css$/,
					use: [
						{
							loader: MiniCssExtractPlugin.loader,
						},
						"css-loader",
					],
				},
			],
		},
		plugins: [
			new MiniCssExtractPlugin({
				filename: "[name].css",
			}),
			new VueLoaderPlugin(),
			new WebpackManifestPlugin({ fileName: 'ssr-manifest.json' })
		]
	}

	return new Promise((resolve, reject) => {
		webpack(config, (err, stats) => {
			if (err) {
				return reject(err)
			}
			if (stats.hasErrors()) {
				return reject(stats.compilation.errors)
			}
			console.log('Build complete.')
			resolve()
		})
	})
}
