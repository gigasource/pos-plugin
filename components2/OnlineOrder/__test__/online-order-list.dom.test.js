import { wrapper, makeWrapper } from "../../test-utils";
import OnlineOrderList from '../OnlineOrderList';

//todo: mock new i18n
describe("test call system ui", () => {
  it("should render", () => {
    makeWrapper(OnlineOrderList);
    expect(wrapper.html()).toMatchInlineSnapshot();
  });
});
