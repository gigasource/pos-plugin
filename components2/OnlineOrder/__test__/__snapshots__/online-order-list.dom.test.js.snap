// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test online order list should render 1`] = `
<div class="online-order-list">
  <div class="row-flex align-items-center mb-2">
    <div class="online-order-list__title">
      onlineOrder.completed onlineOrder.orders
    </div>
    <g-spacer-stub>
    </g-spacer-stub>
    <div class="online-order-list__info">
      <div class="row-flex align-items-center">
        <g-icon-stub class="mr-2"
                     size="20"
        >
          icon-cutleries2
        </g-icon-stub>
        <span class="fw-700">
        </span>
      </div>
      <div class="row-flex align-items-center">
        <g-icon-stub class="mr-2"
                     size="20"
        >
          icon-money-bag
        </g-icon-stub>
        <span class="fw-700">
          'common.currency' , undefined12
        </span>
      </div>
    </div>
    <date-range-picker from="2021-01-14"
                       to="2021-01-14"
    >
    </date-range-picker>
  </div>
  <div class="online-order-list__table">
    <g-table-stub elevation="2"
                  fixed-header="true"
    >
      <thead>
        <tr>
          <th style="white-space: nowrap;">
          </th>
          <th style="white-space: nowrap;">
          </th>
          <th style="white-space: nowrap;">
          </th>
          <th style="white-space: nowrap;">
          </th>
          <th style="white-space: nowrap;">
          </th>
          <th style="white-space: nowrap;">
          </th>
          <th style="white-space: nowrap;">
          </th>
          <th style="white-space: nowrap;">
          </th>
          <th style="white-space: nowrap;">
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td class="fw-700">
            <p style="white-space: nowrap;">
              #1
            </p>
            <g-tooltip openonhover="true"
                       color="#616161"
                       transition="0.3"
                       speech-bubble="true"
                       remove-content-on-close="true"
            >
            </g-tooltip>
          </td>
          <td>
            <p>
              fgh
            </p>
            <p style="white-space: nowrap;">
              0123123131
            </p>
          </td>
          <td>
            <div>
              --
            </div>
          </td>
          <td>
            <p class="fw-700"
               style="white-space: nowrap;"
            >
              'common.currency' , undefined12
            </p>
            <p>
              <img alt="true">
            </p>
          </td>
          <td style="white-space: nowrap;">
            14.01.2021
          </td>
          <td style="white-space: nowrap;">
            15:33
          </td>
          <td>
            <g-icon-stub>
              icon-take-away
            </g-icon-stub>
          </td>
          <td class="completed">
            <div style="white-space: nowrap;">
              onlineOrder.completed
            </div>
            <div style="font-size: x-small; margin-top: -5px;">
            </div>
          </td>
          <td>
          </td>
        </tr>
      </tbody>
    </g-table-stub>
  </div>
  <dialog-complete-order-stub disable-btn="true"
                              modelvalue="false"
  >
  </dialog-complete-order-stub>
</div>
`;
